### ISSUE TITLE
Dockerfile bash injection is too broad

### ISSUE THREAD

**Issue by iyzana** [nvim-treesitter/nvim-treesitter#6530]
### Describe the highlighting problem

The bash injection in `Dockerfile`s is not isolated per `RUN` statement

### Example snippet that causes the problem

```dockerfile
FROM scratch
RUN echo a
RUN echo b
```

### Tree-sitter parsing result

```lisp
(source_file ; [0, 0] - [3, 0] dockerfile
  (program ; [1, 4] - [2, 10] bash
    (command ; [1, 4] - [2, 10] bash
      name: (command_name ; [1, 4] - [1, 8] bash
        (word)) ; [1, 4] - [1, 8] bash
      argument: (word) ; [1, 9] - [2, 8] bash
      argument: (word))) ; [2, 9] - [2, 10] bash
  (from_instruction ; [0, 0] - [0, 12] dockerfile
    "FROM" ; [0, 0] - [0, 4] dockerfile
    (image_spec ; [0, 5] - [0, 12] dockerfile
      name: (image_name))) ; [0, 5] - [0, 12] dockerfile
  "\n" ; [0, 12] - [1, 0] dockerfile
  (run_instruction ; [1, 0] - [1, 10] dockerfile
    "RUN" ; [1, 0] - [1, 3] dockerfile
    (shell_command ; [1, 4] - [1, 10] dockerfile
      (shell_fragment))) ; [1, 4] - [1, 10] dockerfile
  "\n" ; [1, 10] - [2, 0] dockerfile
  (run_instruction ; [2, 0] - [2, 10] dockerfile
    "RUN" ; [2, 0] - [2, 3] dockerfile
    (shell_command ; [2, 4] - [2, 10] dockerfile
      (shell_fragment))) ; [2, 4] - [2, 10] dockerfile
  "\n") ; [2, 10] - [3, 0] dockerfile
```

### Example screenshot

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/16743652/81b42b1a-0731-4012-8801-00360f4b9bfe)

Colorscheme: ellisonleao/gruvbox.nvim

`:Inspect` for `RUN` on the last line:

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/16743652/c7abbd6d-c9fa-431a-a209-2747c2d7ed26)

### Expected behavior

I'd expect there to be two bash injecections, one for each `RUN` statement

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
- OK `node` found v21.7.3 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (GCC) 13.2.1 20240417
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "6.6.28-1-lts",
  sysname = "Linux",
  version = "#1 SMP PREEMPT_DYNAMIC Wed, 17 Apr 2024 10:11:09 +0000"
} ~

Parser/Features         H L F I J
  - bash                ✓ ✓ ✓ . ✓
  - dockerfile          ✓ . . . ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~

The following errors have been detected: ~
- ERROR properties(highlights): /usr/share/nvim/runtime/lua/vim/treesitter/language.lua:112: Failed to load parser for language 'properties': uv_dlopen: no error
  properties(highlights) is concatenated from the following files:
  | [ERROR]:"/home/jannis/.local/share/nvim/plugged/nvim-treesitter/queries/properties/highlights.scm", failed to load: /usr/share/nvim/runtime/lua/vim/treesitter/language.lua:112: Failed to load parser for language 'properties': uv_dlopen: no error
- ERROR properties(locals): /usr/share/nvim/runtime/lua/vim/treesitter/language.lua:112: Failed to load parser for language 'properties': uv_dlopen: no error
  properties(locals) is concatenated from the following files:
  | [ERROR]:"/home/jannis/.local/share/nvim/plugged/nvim-treesitter/queries/properties/locals.scm", failed to load: /usr/share/nvim/runtime/lua/vim/treesitter/language.lua:112: Failed to load parser for language 'properties': uv_dlopen: no error
- ERROR properties(injections): /usr/share/nvim/runtime/lua/vim/treesitter/language.lua:112: Failed to load parser for language 'properties': uv_dlopen: no error
  properties(injections) is concatenated from the following files:
  | [ERROR]:"/home/jannis/.local/share/nvim/plugged/nvim-treesitter/queries/properties/injections.scm", failed to load: /usr/share/nvim/runtime/lua/vim/treesitter/language.lua:112: Failed to load parser for language 'properties': uv_dlopen: no error
```

### Output of `nvim --version`

```text
NVIM v0.10.0-dev-3032+g61063653b0
Build type: RelWithDebInfo
LuaJIT 2.1.1713773202
Run "nvim -V1 -v" for more info
```

### Additional context

Unrelated to this issue: The `Output of :checkhealth nvim-treesitter` section makes it quite hard to search for issues. Because of that section, every highlighting issue mentions a lot of unrelated languages, making it really difficult to narrow down if an issue has been reported already. I reduced my output to the languages relevant for this issue.

**Comment by clason** [comment]
PR welcome; queries are community-maintained.

**Comment by iyzana** [comment]
Okay, I've been reading up on how the queries are written and how they are supposed to behave a bit.
(<https://neovim.io/doc/user/treesitter.html#treesitter-language-injections>)

And if I'm understanding correctly, `injection.combined` does not behave as intended in this case.

The injection query looks like this:

```scm
((shell_command
  (shell_fragment) @injection.content)
  (#set! injection.language "bash")
  (#set! injection.combined))
```

This does not match the `RUN` keyword, which is a child of the `run_instruction` from the `dockerfile`-grammar and not in the `shell_command` tree at all. But the `RUN` token is a part of the combined injection-tree anyway. Concretely: `argument: (word) ; [1, 9] - [2, 8] bash` should not be in the parsing result, since it spans the `RUN` token which was not selected by the query.

That seems to me to be a more general bug than what I reported and I don't know if the code where the injections are combined is part of [treesitter](https://github.com/tree-sitter/tree-sitter) itself or part of nvim-treesitter.

What I also noticed, is that when modifying the Dockerfile example snippet by adding a semicolon behind the first run statement

```
FROM scratch
RUN echo hi;
RUN echo hi
```

and then _removing_ that semicolon again, the highlighting is now correct until the file is reloaded (using `:e` for example) which I believe should never happen.

**Comment by lucario387** [comment]
tldr: We should remove injection.combined

**Comment by faergeek** [comment]
Why do we want `injection.combined` here though? Seems like this works just fine:

```scm
((comment) @injection.content
  (#set! injection.language "comment"))

((shell_command) @injection.content
  (#set! injection.language "bash")
  (#set! injection.include-children))

((run_instruction
  (heredoc_block) @injection.content)
  (#set! injection.language "bash")
  (#set! injection.include-children))
```

So `@injection.content` is on the whole `shell_command` instead of separate fragments. Fragments are separated by `line_continuation` (backslash), which works in bash the same way basically. Not sure if there's something more. And as I understand, `injection.include-children` merges all those fragments together with continuations into a single piece of text, which is then highlighted as `bash`.

**Comment by lucario387** [comment]
Welp, PR welcome then :D

**Comment by faergeek** [comment]
Submitted a PR for that part. I'm also wondering if there's a way to highlight heredoc better. Right now with this source:

```dockerfile
RUN <<EOF
  ls -l
  tree
EOF
```

`<<EOF` at the start and `EOF` at the end are highlighted by bash injections, `<<EOF` is treated as a redirection, `EOF` at the end is treated as a `word`. Is there a way to exclude them from bash injection itself and keep them highlighted by `dockerfile` itself? I couldn't figure that out on my own.

**Comment by iyzana** [comment]
@faergeek That fix works _for now_ but is incorrect and will break for some Dockerfiles.

It works because both bash and Dockerfiles use a backslash for line continuations by default, BUT the escape character for line continuations can be overridden in Dockerfiles using a `# escape=\`` directive.

In that case the \` eoulf be passend in to the bash parser, which will then error.

**Comment by iyzana** [comment]
@lucario387 I believe `injection.combined` is correct here, but I think there's a bug in its implementation

**Comment by Iron-E** [comment]
I played around with this tonight and had some success.

(excuse the poor formatting 🙏)

<details><summary>injections.scm</summary>

```scheme
(
	(comment) @injection.content
	(#set! injection.language "comment")
)

(
	(shell_command
		(
			(shell_fragment) @injection.content
			(line_continuation)?
		)+
	)

	(#set! injection.language "bash")
)

(
	(run_instruction
		(heredoc_block
			(
				(heredoc_line) @injection.content
				"\n" @injection.content
			)+
		)
	)

	(#set! injection.language "bash")
)
```

</details>

| Before | After |
|---|---|
| <img width="726" height="647" alt="Image" src="https://github.com/user-attachments/assets/f78dc75f-6766-48d3-8198-a0785069fe5a" /> | <img width="680" height="661" alt="Image" src="https://github.com/user-attachments/assets/88f2119e-180e-4e51-bcec-8dd529d508fd" /> |

Happy to open a PR if this looks good.
